##################################################
# CONFIGURATION
##################################################
configfile: "config/config.yaml"

resources_dict = {
    "mem_mb": 188000,  # 188 Go
    "disk_mb": 2000000
}

##################################################
# RESOURCES
##################################################

THREADS = config["resources"]["threads"]
MEM = config["resources"]["memory"]
#TIME = config["resources"]["time"]

##################################################
# FUNCTIONS
##################################################
# imports
#from scripts.functions import get_absolute_time


##################################################
# ENVIRONMENT
##################################################
ABRICATE = config["envs"]["abricate"]
FLYE = config["envs"]["flye"]
PORECHOP = config["envs"]["porechop"]
PLOT = config["envs"]["python_plot"]
KRAKEN2 = config["envs"]["kraken2"]
NANOPLOT = config["envs"]["nanoplot"]
BRACKEN = config["envs"]["bracken"]
BCFTOOLS = config["envs"]["bcftools"]
MEDAKA = config["envs"]["medaka"]
HOST_DEP = config["envs"]["host_depletion"]
RGI = config["envs"]["rgi"]
DAG_ENV = config["envs"]["dag"]
MULTIQC = config["envs"]["multiqc"]
PYTHON_FILTER = config["envs"]["python_filter"]
TAXO_ANALYSIS = config["envs"]["taxo_analysis"]
###################################################


############################################################
# PARAMS #
##########
#SEQUENCE_LENGTH = config["params"]["treetime"]["seq_len"]

#ABSOLUTE_TIME = config["params"]["absolute_time"]["date"]
#FILE = config["params"]["absolute_time"]["file"]

SAMPLES, = glob_wildcards("resources/reads/{sample_id}.fastq.gz")

#MODULES = config["modules"]


TAXONOMIC_LEVELS = ["S", "G", "F", "P"]
##################################################


# INCLUDES
##################################################Pre-Traitement et quality
include: "rules/fastplong.smk"
include: "rules/nanoplot.smk"
include: "rules/host_depletion.smk"
include: "rules/rename_reads.smk"
include: "rules/extract_read_stats.smk" 
##################################################Assemblage et Alignment 
include: "rules/flye_assembly.smk"
include: "rules/minimap2_align.smk"
include: "rules/medaka.smk"
include: "rules/bcftools_consensus.smk"
##################################################Annotation AMR
include: "rules/rgi.smk"
include: "rules/rgi_bwt.smk"
include: "rules/rgi_filter.smk"
include: "rules/rgi_select_columns.smk"

##################################################Analyses secondaires & visualisation
include: "rules/r_plot_ARGs_all.smk"

##################################################Reporting
include: "rules/multiqc_nanoplot.smk"
include: "rules/build_dag.smk"

#################################################Taxonomic classification
include: "rules/kraken2.smk"
include: "rules/bracken.smk"
include: "rules/taxonomy.smk" 
#################################################SUMMARY
include: "rules/rgi_aggregate.smk"



rule all:
    input:
        expand("results/{sample}/qc/{sample}_Nanoplot/NanoPlot-report.html", sample=config["samples_id"]),
        expand("results/{sample}/rgi/{sample}.txt", sample=config["samples_id"]), 
        expand("results/{sample}/kraken/{sample}_output.txt", sample=config["samples_id"]),
#        expand("results/{sample}/bracken/{sample}_species.txt", sample=config["samples_id"]),
#        expand("results/{sample}/rgi_bwt", sample=config["samples_id"]),
        expand("results/{sample}/rgi_preprocessing/{sample}_filtered.txt", sample=config["samples_id"]),
        expand("results/{sample}/rgi_preprocessing/{sample}_selected.txt", sample=config["samples_id"]),
        expand("results/{sample}/stats/{sample}_read_stats.txt", sample=config["samples_id"]),
        "results/summary/rgi_all_samples.tsv",
        "results/summary/rgi_aggr_summary.txt",
        "results/stats/sequencing_stats.tsv",
        expand("results/{sample}/bracken/{sample}_{level}.txt",
               sample=config["samples_id"], 
               level=TAXONOMIC_LEVELS),
        expand("results/taxonomy/abundance_matrix_{level}_filtered.tsv", 
               level=TAXONOMIC_LEVELS),
        expand("results/taxonomy/abundance_matrix_{level}_relative.tsv", 
               level=TAXONOMIC_LEVELS),
        "results/taxonomy/alpha_diversity.tsv",
        "results/summary/taxonomy_summary.txt"
